add synthesis plugin docs

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Adds API docs section for synthesis plugins.
Fixes #11675.

UPDATE:

For each relevant high-level-object (e.g., `Clifford`, `LinearFunction` or `PermutationGate`) we are adding a table comparing different plugins available for this object, and the links to the available plugins (to easy finding the relevant documentation).

alexanderivrii committed:
very initial commit

alexanderivrii committed:
fix?

alexanderivrii committed:
temporarily removing

alexanderivrii committed:
is this one adds absolute paths?

alexanderivrii committed:
partially restoring

alexanderivrii committed:
another experiment

alexanderivrii committed:
adding the remaining information

ShellyGarion commented on qiskit/transpiler/passes/synthesis/plugin.py:
```diff
@@ -336,6 +336,161 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
+   :toctree: ../stubs/
+
+   /qiskit/transpiler/passes/synthesis/unitary_synthesis/DefaultUnitarySynthesis
+   /qiskit/transpiler/passes/synthesis/solovay_kitaev_synthesis/SolovayKitaevSynthesis
```
I think that `AQCSynthesisPlugin` also belongs here:
https://github.com/Qiskit/qiskit/blob/main/qiskit/transpiler/passes/synthesis/aqc_plugin.py

ShellyGarion replied:
should these docs replace the docs here: https://docs.quantum.ibm.com/api/qiskit/dev/transpiler_passes#synthesis
as it seems that there is some overlap between them

ElePT committed:
Move docs to corresponding file and link from top.
Remove duplicate AQC plugin entry in synthesis docs and replace with module link.
Remove duplicate transpiler builtin plugin page.

ShellyGarion commented:
I don't think that the file `docs/apidoc/synthesis_aqc.rst` should be removed. It contains the entire AQC synthesis code, and not the `AQCSynthesisPlugin`, so now the following link from `qiskit.synthesis` does not work:
`The Approximate Quantum Compiler is available here: qiskit.synthesis.unitary.aqc`

ElePT commented:
> I don't think that the file `docs/apidoc/synthesis_aqc.rst` should be removed. It contains the entire AQC synthesis code, and not the `AQCSynthesisPlugin`, so now the following link from `qiskit.synthesis` does not work: `The Approximate Quantum Compiler is available here: qiskit.synthesis.unitary.aqc`

@ShellyGarion Let me take a look, I removed the file because sphinx was complaining that it wasn't part of any toctree (because the builtin plugin file was the only one using it), but we should be able to make the link work without having an individual file for it. If I can't find the way, let's bring it back. At least, do you think the synthesis plugin page looks like it was intended to?

ShellyGarion commented:
I think that the `AQCSynthesisPlugin` looks OK, but this page has disappeared:
https://docs.quantum.ibm.com/api/qiskit/synthesis_aqc

ElePT committed:
Generate AQC page

ElePT commented:
@ShellyGarion I think that [cf11cb7](https://github.com/Qiskit/qiskit/pull/11763/commits/cf11cb740e3f136c4440df459dce32dfc7a247e0) has fixed the AQC page. I managed to generate it without an additional rst file.

ElePT committed:
Remove plugin from synthesis docs, refer to new plugin docs.

ShellyGarion commented:
> @ShellyGarion I think that [cf11cb7](https://github.com/Qiskit/qiskit/pull/11763/commits/cf11cb740e3f136c4440df459dce32dfc7a247e0) has fixed the AQC page. I managed to generate it without an additional rst file.

Indeed, it looks much better now. thanks!

![image](https://github.com/Qiskit/qiskit/assets/46566946/216c3f65-209e-4ad1-8e62-d1b24aaac6a9)

ElePT commented on qiskit/transpiler/passes/synthesis/aqc_plugin.py:
```diff
@@ -10,7 +10,14 @@
 # that they have been altered from the originals.
 """
-An AQC synthesis plugin to Qiskit's transpiler.
+==============================================================================
+AQC Synthesis Plugin (in :mod:`qiskit.transpiler.passes.synthesis.aqc_plugin`)
+==============================================================================
```
I tried removing these titles for the aqc/solovay-kitaev/unitary synthesis plugins but I think the result looks a bit strange:

![Screenshot 2024-02-14 at 15 46 08](https://github.com/Qiskit/qiskit/assets/57907331/979f8491-f278-40ad-807e-5c256eda855b)

So I decided to leave the titles, but either way is possible:

![Screenshot 2024-02-14 at 15 47 56](https://github.com/Qiskit/qiskit/assets/57907331/997de78f-f02a-49dd-811a-232f91c885ca)

ElePT commented on qiskit/transpiler/passes/synthesis/high_level_synthesis.py:
```diff
@@ -11,7 +11,127 @@
+   BasicSynthesisPermutation
+   ACGSynthesisPermutation
+   KMSSynthesisPermutation
+   TokenSwapperSynthesisPermutation
```
I also tried integrating these with little success, but I think that having 2 tables is not that bad

![Uploading Screenshot 2024-02-14 at 14.44.53.png…]()

jakelishman commented:
This is marked as closing an issue in the 1.0.0 milestone. Is it ready to review/merge now?

We can always tweak documentation after the release, but it'd be good to have at least the main skeleton in to begin with (and this already looks fairly done?). We could merge this and have more PRs later if that's convenient?

alexanderivrii commented:
Sorry for the delay. I would like to make one more quick pass over the text in the tables. And many thanks to @ElePT for turning the initial mess of this PR to the current form.

alexanderivrii committed:
adding short descriptions of available plugins

jakelishman reviewed (approved):
Thanks for this, Sasha, Elena and Shelly - it looks like a really big improvement over what we've (not!) got so far.
